### PR TITLE
fix(sql-schema-describer): ignore indexes with expressions on postgres

### PR DESCRIPTION
**Pull request by tomhoule**
We previously got rows without corresponding column (indexkeyid = 0),
and just skipped them, causing us to mangle indexes. The result we saw
in introspection CI
(https://prisma-company.slack.com/archives/C01778YKS00/p1657702144980359)
was columns ending up in the _wrong_ index because the first colun in
their original index was an expression. These now all get filtered out.

The first commit is slightly tangential:

feat: add a code-free test setup for introspection-engine-tests

SQL -> introspected prisma schema. We expect better compile times and
fewer diffs due to meaningless changes in test setup after this.

It also makes it easy to add additional checks (introspected schemas are
valid, etc.) without adding or changing calls in tests.

The first test is the reproduction of a regression in indexes
introspection.

### CODE REVIEW

File: libs/datamodel/core/src/validate/common.rs [hunk 1]
```diff
@@ -1,43 +0,0 @@
-use crate::{
-    ast,
-    common::preview_features::*,
-    diagnostics::{DatamodelError, DatamodelWarning, Diagnostics},
-};
-use itertools::Itertools;
-
-pub fn parse_and_validate_preview_features(
```
Comment (tomhoule) [inline libs/datamodel/core/src/validate/common.rs:8]: This just moved. One module fewer. It's only used in one place, so now it's there and it is private.

**Review by pimeys — APPROVED** [review]
No need to zoom. Pretty clear. Maybe start having already database folders in the simple tests, so we don't need to prefix them all and can copy them easily to a new directory when adding a new connector.

File: introspection-engine/introspection-engine-tests/tests/simple/cockroach_index_with_expression_column.sql [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+/*
+now create the indexes with expression columns
+*/
+
+CREATE INDEX index_communication_channels_on_path_and_path_type ON communication_channels (lower((path)::text), path_type);
+CREATE UNIQUE INDEX index_communication_channels_on_user_id_and_path_and_path_type ON communication_channels (user_id, lower((path)::text), path_type);
+CREATE INDEX index_communication_channels_on_confirmation_code ON communication_channels (confirmation_code);
+
+/*
```
[the remark this answers is not in the document]
Comment (tomhoule) [reply introspection-engine/introspection-engine-tests/tests/simple/cockroach_index_with_expression_column.sql:22]: nope, that's the downside of having everything in the same file

File: introspection-engine/introspection-engine-tests/tests/simple/cockroach_index_with_expression_column.sql [hunk 3]
```diff
@@ -0,0 +1,45 @@
+-- tags=CockroachDb
```
[the remark this answers is not in the document]
Comment (tomhoule) [reply introspection-engine/introspection-engine-tests/tests/simple/cockroach_index_with_expression_column.sql]: Yes I wondered about making the directory structure significant too, but I think that may be a bit too much magic. The gold standard for me is [Mitchell Hashimoto's zig compiler test setup](https://github.com/ziglang/zig/issues/11288) so I went with something similar.

[the remark this answers is not in the document]
Comment (tomhoule) [reply introspection-engine/introspection-engine-tests/tests/simple/cockroach_index_with_expression_column.sql]: Anyway it's easy to change later if we want.

[the remark this answers is not in the document]
Comment (pimeys) [reply introspection-engine/introspection-engine-tests/tests/simple/cockroach_index_with_expression_column.sql]: No magic, just spreading the files per database. You'd still have the `-- tags=Foo` in the file, but now it's easy to copy tests to a different database.

[the remark this answers is not in the document]
Comment (tomhoule) [reply introspection-engine/introspection-engine-tests/tests/simple/cockroach_index_with_expression_column.sql]: so the rule would be that the first level below tests/simple/ would automatically be added as a tag? and then we make the `-- tags=...` line optional.

[the remark this answers is not in the document]
Comment (pimeys) [reply introspection-engine/introspection-engine-tests/tests/simple/cockroach_index_with_expression_column.sql]: No we don't need that. We can do it, but I'm allergic to magic... Just that the files are per database, and it's easy to e.g. copy them all to a new connector if needed.

**Comment by tomhoule** [comment]
Ok let's merge and continue the discussion about separating tests by connectors in different directories in the next PRs.

**Pull request merged**